### PR TITLE
FEATURE: QA-18878 Write automation ATCs for Funnel and Pyramid chart on SDK

### PR DESCRIPTION
**Pull request by duclequang**
---

Supported PR commands:

| Command                                                 | Description                                                |
| ------------------------------------------------------- | ---------------------------------------------------------- |
| `ok to test`                                            | Re-run standard checks                                     |
| `extended check sonar`                                  | SonarQube tests                                            |
| `extended test - backstop`                              | BackstopJS tests                                           |
| **E2E Cypress tests commands - TIGER**                  |                                                            |
| `extended test - tiger-cypress - isolated <testName>`   | Run isolated tests running against recorded Tiger backend. |
| `extended test - tiger-cypress - record <testName>`     | Create a new recording for isolated Tiger tests.           |
| `extended test - tiger-cypress - integrated <testName>` | Run integrated tests against live backend                  |
| **E2E Cypress tests commands - BEAR**                   |                                                            |
| `extended test - cypress - isolated <testName>`         | Run isolated tests running against recorded Bear backend.  |
| `extended test - cypress - record <testName>`           | Create a new recording for isolated Bear tests.            |
| `extended test - cypress - integrated <testName>`       | Run integrated tests against live backend                  |
| **Compatibility matrix test commands - TIGER Backend**  |                                                            |
| `extended test - matrix-test <AIO_version>`             | Run integrated tests against AIO versions.                 |

`<testName>` in cypress commands is used to filter specfiles. Example, to run record with BEAR backend

-   Against `dashboard.spec.ts` and `drilling.spec.ts`, execute command `extended test - cypress - record dashboard,drilling`
-   Against all specfiles, execute command `extended test - cypress - record` or `extended test - cypress - record *`

`<AIO_version>` in commands is used to start test with multiple AIO instances - each instance in triggered by one jenkins build

-   To run with `master` and `stable`, execute command `extended test - matrix-test master,stable` or `extended test - matrix-test latest`
-   To run with specific version,ex: `2.3.0` and `2.3.1`, execute command `extended test - matrix-test 2.3.0,2.3.1`
-   In case `<AIO_version>` is empty, read versions from file `compTigerVersions.txt` of this repo

---

# PR Checklist

-   [ ] commit messages adhere to the [commit message guidelines](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#what-should-the-commits-look-like)
-   [ ] review was done by a Code owner [if necessary](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-tell-if-my-pull-request-needs-approval-by-a-code-owner) (if you think it is not necessary, explain the reasoning in the description or in a comment)
-   [ ] `check` passes
-   [ ] `extended test - backstop` passes
-   [ ] `extended test - tiger-cypress - record` to record new mapping files (Tiger BE)
-   [ ] `extended test - cypress - record` to record new mapping files (Bear BE)
-   [ ] `extended test - tiger-cypress - isolated` passes
-   [ ] `extended test - cypress - isolated` passes
-   [ ] `extended test - tiger-cypress - integrated` passes
-   [ ] `rush change` [was run if applicable](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-describe-my-changes-for-the-changelog)

### CODE REVIEW

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/13590/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/15243/ : SUCCESS in 8m 15s

**Comment by duclequang** [comment]
extended test - tiger-cypress - integrated

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/13591/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-pull-request-dispatcher-pipeline/747/ : SUCCESS in 10m 13s

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/13592/ : SUCCESS in 5s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-pull-request-dispatcher-pipeline/748/ : FAILURE in 7m 54s

**Comment by duclequang** [comment]
extended test - cypress - integrated

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/13593/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-pull-request-dispatcher-pipeline/749/ : FAILURE in 7m 20s

File: libs/sdk-ui-tests-e2e/cypress/integration/01-sdk-ui-dashboard/dashboardTigerWithCharts.spec.ts [hunk 1]
```diff
@@ -28,3 +28,23 @@ describe("Dashboard with charts", { tags: ["pre-merge_isolated_tiger"] }, () =>
         });
     });
 });
+
+describe(
+    "Dashboard with pyramid and funnel charts",
+    { tags: ["checklist_integrated_tiger", "checklist_integrated_tiger"] },
```
Comment (duongthimy) [inline libs/sdk-ui-tests-e2e/cypress/integration/01-sdk-ui-dashboard/dashboardTigerWithCharts.spec.ts:34]: Is this a wrong about the typo tag name

File: libs/sdk-ui-tests-e2e/cypress/integration/03-sdk-ui-charts/pyramidChart.spec.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
+// (C) 2021 GoodData Corporation
+import * as Navigation from "../../tools/navigation";
+import { Chart } from "../../tools/chart";
+
+describe("Pyramid Chart", { tags: ["checklist_integrated_tiger", "checklist_integrated_bear"] }, () => {
```
Comment (duongthimy) [inline libs/sdk-ui-tests-e2e/cypress/integration/03-sdk-ui-charts/pyramidChart.spec.ts:5]: we don't need to run this test for 2 backends, only for tiger is enough

File: libs/sdk-ui-tests-e2e/scenarios/src/components/Scenarios/Dashboard/DashboardScenarioTigerCharts.tsx [hunk 3]
```diff
@@ -2,8 +2,16 @@
...
-import { Dashboards } from "../../../../../reference_workspace/workspace_objects/goodsales/current_reference_workspace_objects_tiger";
+
+import * as BearMDObjects from "../../../../../reference_workspace/workspace_objects/goodsales/current_reference_workspace_objects_bear";
+import * as TigerMDObjects from "../../../../../reference_workspace/workspace_objects/goodsales/current_reference_workspace_objects_tiger";
+
+type MDObjectsType = typeof TigerMDObjects & typeof BearMDObjects;
+
+export const MDObjects = (
+    process.env.SDK_BACKEND === "TIGER" ? TigerMDObjects : BearMDObjects
+) as MDObjectsType;
```
Comment (duongthimy) [inline libs/sdk-ui-tests-e2e/scenarios/src/components/Scenarios/Dashboard/DashboardScenarioTigerCharts.tsx:13]: As the previous comment, we don't need to run UI SDK test for both of backend, so we don't need to IF, ELSE here to select the MDObject

File: libs/sdk-ui-tests-e2e/scenarios/src/components/Scenarios/Visualizations/FunnelChart/FunnelChartScenario.tsx [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+import React from "react";
+import { FunnelChart } from "@gooddata/sdk-ui-charts";
+import * as BearMDObjects from "../../../../../../reference_workspace/workspace_objects/goodsales/current_reference_workspace_objects_bear";
+import * as TigerMDObjects from "../../../../../../reference_workspace/workspace_objects/goodsales/current_reference_workspace_objects_tiger";
+
+type MDObjectsType = typeof TigerMDObjects & typeof BearMDObjects;
+
+export const MDObjects = (
+    process.env.SDK_BACKEND === "TIGER" ? TigerMDObjects : BearMDObjects
+) as MDObjectsType;
```
Comment (duongthimy) [inline libs/sdk-ui-tests-e2e/scenarios/src/components/Scenarios/Visualizations/FunnelChart/FunnelChartScenario.tsx:11]: As the previous comment, we don't need to run UI SDK test for both of backend, so we don't need to IF, ELSE here to select the MDObject

File: libs/sdk-ui-tests-e2e/scenarios/src/components/Scenarios/Visualizations/PyramidChart/PyramidChartScenario.tsx [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+import React from "react";
+import { PyramidChart } from "@gooddata/sdk-ui-charts";
+import * as BearMDObjects from "../../../../../../reference_workspace/workspace_objects/goodsales/current_reference_workspace_objects_bear";
+import * as TigerMDObjects from "../../../../../../reference_workspace/workspace_objects/goodsales/current_reference_workspace_objects_tiger";
+
+type MDObjectsType = typeof TigerMDObjects & typeof BearMDObjects;
+
+export const MDObjects = (
+    process.env.SDK_BACKEND === "TIGER" ? TigerMDObjects : BearMDObjects
+) as MDObjectsType;
```
Comment (duongthimy) [inline libs/sdk-ui-tests-e2e/scenarios/src/components/Scenarios/Visualizations/PyramidChart/PyramidChartScenario.tsx:11]: As the previous comment, we don't need to run UI SDK test for both of backend, so we don't need to IF, ELSE here to select the MDObject

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by duclequang** [comment]
extended test - tiger-cypress - integrated

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/13597/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/15247/ : FAILURE in 8m 34s

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/13598/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-pull-request-dispatcher-pipeline/751/ : SUCCESS in 9m 27s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/13601/ : SUCCESS in 5s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/15248/ : SUCCESS in 10m 08s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/13612/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/15258/ : SUCCESS in 7m 43s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/13616/ : SUCCESS in 5s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/15262/ : SUCCESS in 8m 20s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/13627/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/15270/ : SUCCESS in 7m 49s

File: libs/sdk-ui-tests-e2e/scenarios/src/components/Scenarios/Visualizations/FunnelChart/FunnelChartScenario.tsx [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+    Amount,
+    Department,
+} from "../../../../../../reference_workspace/workspace_objects/goodsales/current_reference_workspace_objects_tiger";
+
+const style = { height: 300 };
+
+export const FunnelChartScenario: React.FC = () => {
+    return (
+        <div style={style} className="s-funnel-chart">
+            <FunnelChart measures={[Amount, Department]} />
```
Comment (duongthimy) [inline libs/sdk-ui-tests-e2e/scenarios/src/components/Scenarios/Visualizations/FunnelChart/FunnelChartScenario.tsx:14]: We should create a Funnel chart with 1 measure + 1 attribute, not 2 measures like above

File: libs/sdk-ui-tests-e2e/scenarios/src/components/Scenarios/Visualizations/PyramidChart/PyramidChartScenario.tsx [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+    Amount,
+    Department,
+} from "../../../../../../reference_workspace/workspace_objects/goodsales/current_reference_workspace_objects_tiger";
+
+const style = { height: 300 };
+
+export const PyramidChartScenario: React.FC = () => {
+    return (
+        <div style={style} className="s-pyramid-chart">
+            <PyramidChart measures={[Amount, Department]} />
```
Comment (duongthimy) [inline libs/sdk-ui-tests-e2e/scenarios/src/components/Scenarios/Visualizations/PyramidChart/PyramidChartScenario.tsx:14]: We should create a Pyramid chart with 1 measure + 1 attribute, not 2 measures like above

File: libs/sdk-ui-tests-e2e/cypress/integration/01-sdk-ui-dashboard/dashboardTigerWithCharts.spec.ts [hunk 8]
```diff
@@ -28,3 +28,19 @@ describe("Dashboard with charts", { tags: ["pre-merge_isolated_tiger"] }, () =>
         });
     });
 });
+
+describe("Dashboard with pyramid and funnel charts", { tags: ["checklist_integrated_tiger"] }, () => {
+    beforeEach(() => {
+        Navigation.visit("dashboard/dashboard-tiger-charts");
+    });
+
+    it("should render funnel and pyramid charts correctly", () => {
```
Comment (duongthimy) [inline libs/sdk-ui-tests-e2e/cypress/integration/01-sdk-ui-dashboard/dashboardTigerWithCharts.spec.ts:37]: should rename to "should render default color legend of funnel and pyramid chart correctly"

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/13631/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/15275/ : SUCCESS in 7m 45s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/13632/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/15278/ : FAILURE in 8m 24s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by duclequang** [comment]
https://checklist.intgdc.com/job/cypress-tests/job/gooddata-ui-sdk-cypress-checklist_integrated_tiger/437/ >> pyramidChart.spec.ts
https://checklist.intgdc.com/job/cypress-tests/job/gooddata-ui-sdk-cypress-checklist_integrated_tiger/435/ >> dashboardTigerWithCharts.spec.ts
https://checklist.intgdc.com/job/cypress-tests/job/gooddata-ui-sdk-cypress-checklist_integrated_tiger/436/ >> funnelChart.spec.ts

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/13641/ : SUCCESS in 5s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/15292/ : SUCCESS in 7m 44s

**Review by duongthimy — APPROVED** [review]

**Comment by duongthimy** [comment]
LGTM

**Comment by yenkins** [comment]
### Sonar scan result


More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ui-sdk-gate-PR3768

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-sonarqube-scan-zuul-gate/2738/ : SUCCESS in 9m 08s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/15295/ : SUCCESS in 7m 41s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-cypress-bear-isolated-zuul-pipeline/272/ : SUCCESS in 10m 43s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-merge-compliance-check/3895/ : SUCCESS in 9s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-fossa-scanning-zuul/620/ : FAILURE in 1m 13s (non-voting)
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-cypress-tiger-isolated-zuul-pipeline/241/ : SUCCESS in 9m 52s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-compatibility-pipeline/126/ : SUCCESS in 4m 29s (non-voting)

**Pull request merged**